src/uct/cuda/cuda_copy/cuda_copy_ep.c

```diff
@@ -187,7 +276,12 @@ uct_cuda_copy_post_cuda_async_copy(uct_ep_h tl_ep, void *dst, void *src,
     ucs_trace("cuda async issued: %p dst:%p[%s], src:%p[%s] len:%ld",
               cuda_event, dst, ucs_memory_type_names[dst_type], src,
               ucs_memory_type_names[src_type], length);
-    return UCS_INPROGRESS;
+    status = UCS_INPROGRESS;
+
+out_pop_and_release:
+    uct_cuda_primary_ctx_pop_and_release(cuda_device);
```

Comment:
seems it is risky to release, because we are still using this context resources. If user also releases the primary context, all related resources (stream, events, etc) will be also destroyed before the operation completes.
see description in #10578

Reply:
Here we retain and push only the active context. This means that the context was activated before this call. I'm not sure that we shouldn't release it here.

Reply:
Also it will not help if user calls `cuDevicePrimaryCtxReset`, or `cudaDeviceReset`. It only works with user's `cuDevicePrimaryCtxRelease` call.

Moreover, we check that the context is valid before destroying the resources. And we report errors/warnings in case of failed CUDA Driver API calls. This should be enough for these cases. Maybe we can review event/stream usage and enhance error flows, and enhance unit tests by covering these corner cases.